### PR TITLE
Reapply "[RISCV] Support RISCV Atomics ABI attributes (#84597)"

### PR DESCRIPTION
**Pull request by ilovepi**
This patch adds support for the atomic_abi attribute, specifid in
https://github.com/riscv-non-isa/riscv-elf-psabi-doc/blob/master/riscv-elf.adoc#tag_riscv_atomic_abi-14-uleb128version.

This was previously reverted due to ld.bfd segfaulting w/ unknown riscv
attributes. Attribute emission is now guarded by a backend flag
`--riscv-abi-attributes`, which is off by default. Linker support in
LLD for attribute merging is now in a standalone patch.

### CODE REVIEW

**Comment by ilovepi** [comment]
Note, I'm not sure what the time table for this to reland is.

**Comment by ilovepi** [comment]
Seems like @MaskRay reported the bfd bug quite a while ago. https://sourceware.org/bugzilla/show_bug.cgi?id=29823

**Comment by MaskRay** [comment]
> Seems like @MaskRay reported the bfd bug quite a while ago. [sourceware.org/bugzilla/show_bug.cgi?id=29823](https://sourceware.org/bugzilla/show_bug.cgi?id=29823)

Hopefully @Nelson1225 can take a look.

---

As we add more build attributes, `llvm-readelf -A riscv64.o` starts to resemble a typical AArch32 relocatable file with over 10 attributes.
This might not be ideal for everyone, and I haven't found another architecture doing this.

The GNU ld error suggests the atomic attribute should be opt-in.
This allows groups who find these attributes useful to enable them explicitly.

On the LLVM side, we can have a `cl::opt<bool>` option to control whether such attributes should be emitted.
We can replace this with a driver option after we gain more experience and agree with GCC.
It is possible that we will eventually go the AArch32 way and a typical relocatable file will contain 10+ attributes, but that's probably not the immediate focus.

File: lld/ELF/Arch/RISCV.cpp
Comment (MaskRay) [inline lld/ELF/Arch/RISCV.cpp:1194]: no braces for one-line single statements

see other comments in #84597

**Review by MaskRay — CHANGES_REQUESTED** [review]
.

**Comment by ilovepi** [comment]
> > Seems like @MaskRay reported the bfd bug quite a while ago. [sourceware.org/bugzilla/show_bug.cgi?id=29823](https://sourceware.org/bugzilla/show_bug.cgi?id=29823)
> 
> Hopefully @Nelson1225 can take a look.
> 
> As we add more build attributes, `llvm-readelf -A riscv64.o` starts to resemble a typical AArch32 relocatable file with over 10 attributes. This might not be ideal for everyone, and I haven't found another architecture doing this.
> 
> The GNU ld error suggests the atomic attribute should be opt-in. This allows groups who find these attributes useful to enable them explicitly.
> 

I'm not sure I follow, are you refering to the current ToT GNU ld behavior or the segfault reported? Second, the psABI isn't ambiguous about the requirement here https://github.com/riscv-non-isa/riscv-elf-psabi-doc/blob/master/riscv-elf.adoc#tag_riscv_atomic_abi-14-uleb128version, and the merge policy is clearly spelled out
```
Merge policy
 The linker should report errors if object files with incompatible atomics ABIs are merged; the compatibility rules for atomic ABIs can be found in the compatibility column in the following table.
```
The wording there doesn't sound optional to me.

> On the LLVM side, we can have a `cl::opt<bool>` option to control whether such attributes should be emitted. 

I don't think this is very practical. CL opts are probably not enough, and tend to cause issues, particularly when they don't get forwarded to the linker correctly in LTO builds. 

>We can replace this with a driver option after we gain more experience and agree with GCC. It is possible that we will eventually go the AArch32 way and a typical relocatable file will contain 10+ attributes, but that's probably not the immediate focus.

I'm not aware of any reason we need to maintain strict compatibility w/ GCC at all points in time, when we're all supposed to be implementing the same spec. Note: I'm not saying we shouldn't maintain compatibility at all, but I don't see why Clang can't be leading the way here. Also, whether we emit attributes or not, doesn't seem to be the kind of thing you should control w/ a driver option if its mandated by the ABI you're targeting.

Lastly, I'd like to emphasize, enabling the correct Atomics ABI has been blocked on this feature for a long time. This is required for Android and other systems that do not want to maintain **any** compatibility with the old/broken A6C ABI, since A6C and A7 are completely incompatible.

File: lld/ELF/Arch/RISCV.cpp
Comment (ilovepi) [reply lld/ELF/Arch/RISCV.cpp:1194]: Ah I missed that. Thanks.

**Comment by asb** [comment]
> I don't think this is very practical. CL opts are probably not enough, and tend to cause issues, particularly when they don't get forwarded to the linker correctly in LTO builds.

I think one reason we'd go with this route is it would make it easy for people to have a downstream patch that flips the default, while having the logic upstream and at least minimally tested. Not ideal, but it's a pattern we've followed before.

Anyway, let's discuss a path forwards in the call today. I agree we could go overboard on maintaining strict compatibility at all times, but the segfault with what was until recently the most current ld release is fairly user hostile. There aren't any great solutions here I think, we just have to find the least-worst :/

**Comment by ilovepi** [comment]
Ping. Are we happy w/ this as an `-mattr`, or should it be guarded by another kind of flag?

**Comment by asb** [comment]
> Ping. Are we happy w/ this as an `-mattr`, or should it be guarded by another kind of flag?

Sorry, I'd missed the update somehow. I'd actually been thinking a cl::opt flag (much like AArch64 has aarch64-mark-bti-property). I'd say the `+abi-attr`/`FeatureAbiAttributes` is confusingly named as well, as it controls only the atomics ABI attributes rather than all of them.

Unless anyone else feels differently I'd think the cl::opt is most consistent with how we usually do this (and easy to pass through from clang with `-mllvm`). The precise name doesn't matter much, as long as it's clear it relates to atomic abi attribute / tags in some way I promise I won't bikeshed beyond that!

**Comment by ilovepi** [comment]
> > Ping. Are we happy w/ this as an `-mattr`, or should it be guarded by another kind of flag?
> 
> Sorry, I'd missed the update somehow. I'd actually been thinking a cl::opt flag (much like AArch64 has aarch64-mark-bti-property). I'd say the `+abi-attr`/`FeatureAbiAttributes` is confusingly named as well, as it controls only the atomics ABI attributes rather than all of them.
> 
> Unless anyone else feels differently I'd think the cl::opt is most consistent with how we usually do this (and easy to pass through from clang with `-mllvm`). The precise name doesn't matter much, as long as it's clear it relates to atomic abi attribute / tags in some way I promise I won't bikeshed beyond that!

Sounds good. I’ll try to send an update later today.

**Comment by MaskRay** [comment]
> > > Seems like @MaskRay reported the bfd bug quite a while ago. [sourceware.org/bugzilla/show_bug.cgi?id=29823](https://sourceware.org/bugzilla/show_bug.cgi?id=29823)
> > 
> > Hopefully @Nelson1225 can take a look.
> > 
> > As we add more build attributes, `llvm-readelf -A riscv64.o` starts to resemble a typical AArch32 relocatable file with over 10 attributes. This might not be ideal for everyone, and I haven't found another architecture doing this.
> > 
> > The GNU ld error suggests the atomic attribute should be opt-in. This allows groups who find these attributes useful to enable them explicitly.
> > 
> 
> I'm not sure I follow, are you refering to the current ToT GNU ld behavior or the segfault reported? Second, the psABI isn't ambiguous about the requirement here https://github.com/riscv-non-isa/riscv-elf-psabi-doc/blob/master/riscv-elf.adoc#tag_riscv_atomic_abi-14-uleb128version, and the merge policy is clearly spelled out

The trend of liberally including attributes in .riscv.attributes raises concerns.
While specific attributes can be helpful, overabundance similar to aarch32's style could lead to clutter and maintenance challenges.
Removing now-unnecessary attributes might become difficult due to some users favoring stronger compatibility.

I'll likely be more concerned if a future linker option is proposed to suppress incompatibility errors.

> ```
> Merge policy
>  The linker should report errors if object files with incompatible atomics ABIs are merged; the compatibility rules for atomic ABIs can be found in the compatibility column in the following table.
> ```
> The wording there doesn't sound optional to me.

I don't think we should read too much from the use of "should" here. This requirement level is probably between "should" and "may".
It should certainly be fine for a linker to ignore `.riscv.attributes` completely.

RFC2119
```
3. SHOULD   This word, or the adjective "RECOMMENDED", mean that there
   may exist valid reasons in particular circumstances to ignore a
   particular item, but the full implications must be understood and
   carefully weighed before choosing a different course.
```

> > On the LLVM side, we can have a `cl::opt<bool>` option to control whether such attributes should be emitted. 
> 
> I don't think this is very practical. CL opts are probably not enough, and tend to cause issues, particularly when they don't get forwarded to the linker correctly in LTO builds. 

Using `cl::opt<bool>` looks good to me.
The LTO status is not a practical issue. Many codegen options are not recorded in LLVM IR, and it's possibly inpractical to record every one (a lot don't even have clear merge policy).

> >We can replace this with a driver option after we gain more experience and agree with GCC. It is possible that we will eventually go the AArch32 way and a typical relocatable file will contain 10+ attributes, but that's probably not the immediate focus.
> 
> I'm not aware of any reason we need to maintain strict compatibility w/ GCC at all points in time, when we're all supposed to be implementing the same spec. Note: I'm not saying we shouldn't maintain compatibility at all, but I don't see why Clang can't be leading the way here. Also, whether we emit attributes or not, doesn't seem to be the kind of thing you should control w/ a driver option if its mandated by the ABI you're targeting.
> 
> Lastly, I'd like to emphasize, enabling the correct Atomics ABI has been blocked on this feature for a long time. This is required for Android and other systems that do not want to maintain **any** compatibility with the old/broken A6C ABI, since A6C and A7 are completely incompatible.

A `cl::opt` looks good to me. By using an internal option `cl::opt<bool>`, we don't commit to the stability of the option.
It could be named, though we probably won't without a good reason. Users need to be ready to change.

**Comment by ilovepi** [comment]
Sorry for the delay on this. I lost track of this w/ other priorities. I think the new implementation does what everyone wants it to do: guard codegen w/ the `cl::opt`. It's less clear to me what we'll do w/ enabling trailing fence by default, w/o the attributes on by default, but that's orthogonal to this.

**Comment by preames** [comment]
> Sorry for the delay on this. I lost track of this w/ other priorities. I think the new implementation does what everyone wants it to do: guard codegen w/ the `cl::opt`. 

I'm also coming into this late, triggered by the attempt to make TSO non-experimental.  This is all interconnected.  Having the attribute emission off by default under a CL opt seems like a reasonable stepping stone.  We could maybe be fancy here in the future and do linker version sniffing, but I'm 100% in support of starting simple.  LGTM to that bit, and assuming other comments on resolved from previous landing - I did not check - please reland.  

> It's less clear to me what we'll do w/ enabling trailing fence by default, w/o the attributes on by default, but that's orthogonal to this.

I think we should enable the A6S ("A6/7 compat") by default, even if we can't emit the attribute.  The whole attribute mechanism is fairly badly thought out - as the very revert here shows - and should not block the major item of getting us migrated to the same ABI variant used by gcc.

**Comment by ilovepi** [comment]
> > Sorry for the delay on this. I lost track of this w/ other priorities. I think the new implementation does what everyone wants it to do: guard codegen w/ the `cl::opt`.
> 
> I'm also coming into this late, triggered by the attempt to make TSO non-experimental. This is all interconnected. Having the attribute emission off by default under a CL opt seems like a reasonable stepping stone. We could maybe be fancy here in the future and do linker version sniffing, but I'm 100% in support of starting simple. LGTM to that bit, and assuming other comments on resolved from previous landing - I did not check - please reland.
> 
> > It's less clear to me what we'll do w/ enabling trailing fence by default, w/o the attributes on by default, but that's orthogonal to this.
> 
> I think we should enable the A6S ("A6/7 compat") by default, even if we can't emit the attribute. The whole attribute mechanism is fairly badly thought out - as the very revert here shows - and should not block the major item of getting us migrated to the same ABI variant used by gcc.

Once this lands, I don't think there are any blockers to #90267, which enables A6S. I've wanted that to be te default atomic ABI for a long time now.

**Comment by ilovepi** [comment]
@MaskRay does the current implementation address your remaining concerns? If so, then I'll go ahead and land this to unblock the TSO work.

File: lld/ELF/Arch/RISCV.cpp
Comment (MaskRay) [inline lld/ELF/Arch/RISCV.cpp:1106]: We have similar diagnostics in two places. One is OutputSections.cpp:127. Follow its style?

File: lld/ELF/Arch/RISCV.cpp
Comment (MaskRay) [inline lld/ELF/Arch/RISCV.cpp:1128]: This introduces some duplicate. Extract to a lambda

File: llvm/lib/Target/RISCV/MCTargetDesc/RISCVTargetStreamer.cpp
Comment (MaskRay) [inline llvm/lib/Target/RISCV/MCTargetDesc/RISCVTargetStreamer.cpp:29]: remove `cl::init(false)`

File: llvm/include/llvm/Support/RISCVAttributes.h
Comment (MaskRay) [inline llvm/include/llvm/Support/RISCVAttributes.h:38]: scoped enum (`enum class`) is better than namespace+unscoped enum

File: lld/ELF/Arch/RISCV.cpp
Comment (ilovepi) [reply lld/ELF/Arch/RISCV.cpp:1106]: hmm, I followed the diagnostic style for STACK_ALIGN originally, but I've kind of split the difference between the OutputSections diagnostic and this one. Happy to amend again.

File: llvm/lib/Target/RISCV/MCTargetDesc/RISCVTargetStreamer.cpp
Comment (ilovepi) [reply llvm/lib/Target/RISCV/MCTargetDesc/RISCVTargetStreamer.cpp:29]: done

File: llvm/include/llvm/Support/RISCVAttributes.h
Comment (ilovepi) [reply llvm/include/llvm/Support/RISCVAttributes.h:38]: I've followed your advice here, but this seems to require quite a lot of `static_casting` when what we actually want is a typed alias for an integer which is precisely what a standard enum type is. I've also had to introduce what I'd say are useless cases to the switch for unreachable types we've already checked.

**Comment by preames** [comment]
Since we're going to have the flag emission under a flag anyways, can we separate the LLD part into a separate review?  Doing so would seem to unblock the compiler side changes, and let the LLD review comments move at their own pace.

**Comment by ilovepi** [comment]
> Since we're going to have the flag emission under a flag anyways, can we separate the LLD part into a separate review? Doing so would seem to unblock the compiler side changes, and let the LLD review comments move at their own pace.

Done. LLD support is in https://github.com/llvm/llvm-project/pull/97347

File: llvm/lib/Target/RISCV/MCTargetDesc/RISCVTargetStreamer.cpp
Comment (MaskRay) [inline llvm/lib/Target/RISCV/MCTargetDesc/RISCVTargetStreamer.cpp:29]: remove `cl::init(false),` the convention is to omit default false.

**Review by MaskRay — APPROVED** [review]

**Review by MaskRay — APPROVED** [review]

File: llvm/lib/Target/RISCV/MCTargetDesc/RISCVTargetStreamer.cpp
Comment (ilovepi) [reply llvm/lib/Target/RISCV/MCTargetDesc/RISCVTargetStreamer.cpp:29]: ugh, I forgot to fix that, fixed up a bad merge. Thanks for noticing that.

**Pull request merged**
